### PR TITLE
WIP - Fixes #2000: Add bulk delete and regroup actions in RestGrids

### LINKED ISSUE #2000: Support bulk actions on Admin configs + prefs
I would like to support the following two actions as context-menu options on the Configs, Prefs, and User Prefs admin grids:

- Delete - will delete all selected entries, with a prompt of course.
- Set Group - will prompt user to enter a new group, and re-categorize all selected records accordingly (this one N/A for User Prefs)

I realize this might get into some work on RestGrid - notably for the delete option - although I would be in favor of supporting this as a custom action in these admin grids if that meant we avoided blocking on a larger change. I think it's worth it to allow us to better clean up and maintain these records, which we have identified as a priority for us across our apps.

### PR DESCRIPTION
**Pull request by febbraiod**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [x] Added CHANGELOG entry, or determined not required.
- [x] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [x] Updated doc comments / prop-types, or determined not required.
- [x] Reviewed and tested on Mobile, or determined not required.
- [x] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

**Comment by febbraiod** [comment]
Requires: https://github.com/xh/hoist-core/pull/161

File: desktop/cmp/rest/RestGridModel.js [hunk 1]
```diff
@@ -174,60 +192,71 @@ export class RestGridModel {
...
+
+        XH.toast({intent, message});
     }
 
-    @action
     editRecord(record) {
         this.formModel.openEdit(record);
     }
 
-    @action
```
Comment (febbraiod) [inline desktop/cmp/rest/RestGridModel.js:202]: None of these "actions" appear to be modifying any observables... maybe left over decorators from a time when they did? Please correct if wrong

File: desktop/cmp/rest/RestGridModel.js [hunk 2]
```diff
@@ -174,60 +192,71 @@ export class RestGridModel {
         this.formModel.openClone(clone);
     }
 
-    @action
-    deleteSelection() {
```
Comment (febbraiod) [inline desktop/cmp/rest/RestGridModel.js:178]: As mentioned in Changelog, I did not find uses of this or editSelection. Assuming that RestGrids aren't used too extensively in client app, so while this hypothetically could be a breaking change, more likely it's just cruft

File: CHANGELOG.md [hunk 3]
```diff
@@ -16,6 +18,13 @@
 * Resetting Grid columns to their default state (e.g. via the Column Chooser) retains enhancements
   applied from matching Store fields.
 
+### 💥 Breaking Changes
+
+* The `deleteSelection` and `editSelection` methods have been removed from `RestGridModel`. These
+  methods are no longer called internally by Hoist. This could be breaking change in the unlikely
+  event that an application is relying on them.
```
Comment (febbraiod) [inline CHANGELOG.md:25]: Remove

File: admin/tabs/preferences/PreferenceModel.js [hunk 4]
```diff
@@ -76,6 +83,17 @@ export class PreferenceModel {
         actionWarning: {
             del: 'Are you sure you want to delete? Deleting preferences can break running apps.'
```
Comment (febbraiod) [inline admin/tabs/preferences/PreferenceModel.js:84]: Use closure here to warn about the number of prefs being deleted

File: desktop/cmp/rest/RestGridModel.js [hunk 5]
```diff
@@ -55,6 +56,20 @@ export const deleteAction = {
     actionFn: ({record, gridModel}) => gridModel.restGridModel.confirmDeleteRecord(record)
 };
 
+export const bulkDeleteAction = {
+    text: 'Delete',
+    icon: Icon.delete(),
+    intent: 'danger',
+    recordsRequired: true,
+    actionFn: ({gridModel}) => gridModel.restGridModel.confirmBulkDeleteRecords()
+};
```
Comment (febbraiod) [inline desktop/cmp/rest/RestGridModel.js:65]: Could use this everywhere we want delete...rather than keeping old single del around. Then bulk is driven by multi select enabled or not. Is this too dangerous?

File: desktop/cmp/rest/RestGridModel.js [hunk 6]
```diff
@@ -174,60 +192,71 @@ export class RestGridModel {
...
     }
 
-    @action
-    deleteSelection() {
-        const record = this.selectedRecord;
-        if (record) this.deleteRecord(record);
-    }
-
-    @action
     deleteRecord(record) {
```
Comment (febbraiod) [inline desktop/cmp/rest/RestGridModel.js:195]: properly mark as async

File: desktop/cmp/rest/RestGridModel.js [hunk 7]
```diff
@@ -174,60 +192,71 @@ export class RestGridModel {
...
     viewRecord(record) {
         this.formModel.openView(record);
     }
 
-    confirmDeleteSelection() {
-        const record = this.selectedRecord;
-        if (record) this.confirmDeleteRecord(record);
-    }
-
     confirmDeleteRecord(record) {
```
Comment (febbraiod) [inline desktop/cmp/rest/RestGridModel.js:219]: rename methods that do this commit... confirmAndDeleteRecord (?)

File: desktop/cmp/rest/RestGridModel.js [hunk 8]
```diff
@@ -174,60 +192,71 @@ export class RestGridModel {
...
+
+    confirmBulkDeleteRecords() {
+        const records = this.selection,
+            warning = this.actionWarning.del;
+
+        if (!warning) {
+            this.bulkDeleteRecordsAsync(records);
         } else {
-            this.deleteRecord(record);
+            const message = isFunction(warning) ? warning(records) : warning;
```
Comment (febbraiod) [inline desktop/cmp/rest/RestGridModel.js:242]: Make sure the edits and add warning know about the function form.

File: desktop/cmp/rest/data/RestStore.js [hunk 9]
```diff
@@ -63,6 +77,19 @@ export class RestStore extends UrlStore {
...
+    async regroupRecordsAsync(ids, groupName) {
+        const {url} = this,
+            resp = await XH.fetchService.putJson({
+                url: `${url}/bulkUpdate`,
+                body: {ids, newParams: {groupName}}
+            }).linkTo(
+                this.loadModel
+            );
+
+        this.loadAsync();
```
Comment (febbraiod) [inline desktop/cmp/rest/data/RestStore.js:89]: await this

File: desktop/cmp/rest/data/RestStore.js [hunk 10]
```diff
@@ -53,6 +53,20 @@ export class RestStore extends UrlStore {
...
+        const {url} = this,
+            ids = records.map(it => it.id),
+            resp = await XH.fetchJson({
+                url: `${url}/bulkDelete`,
+                params: {ids}
+            }).linkTo(
+                this.loadModel
+            );
+
+        this.loadAsync();
```
Comment (febbraiod) [inline desktop/cmp/rest/data/RestStore.js:66]: await this

File: desktop/cmp/rest/regroup/RegroupDialogModel.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+ *
+ * Copyright © 2020 Extremely Heavy Industries Inc.
+ */
+import {uniq} from 'lodash';
+import {HoistModel, XH} from '@xh/hoist/core';
+import {action, bindable, observable} from '@xh/hoist/mobx';
+
+
+@HoistModel
+export class RegroupDialogModel {
```
Comment (febbraiod) [inline desktop/cmp/rest/regroup/RegroupDialogModel.js:13]: Regroup is very specific to configs and prefs. This change should be in a separate PR and apply to only Config and Pref rather than included in the rest package and the action apply to rest grid.

**Comment by lbwexler** [comment]
Don and I reviewed this in detail.  He will break this up into multiple PRs -- deletion support directly in the core component, regrouping support for configs and prefs only.

**Comment by febbraiod** [comment]
Leaving PR open to preserve comment. Will close and delete branch once both features are properly split out
